File: frontend/Business/tg_bot/ApiClient.py
import httpx
import json
from threading import Lock

from Config import *


class ApiClient:
    def __init__(self, obj_class):
        self.obj_class = obj_class
        self.endpoint = obj_class.__name__.lower()
        self.lock = Lock()

    def get_url(self, id=None):
        if id is None:
            return f"{api_base_url}/api/{self.endpoint}"
        else:
            return f"{api_base_url}/api/{self.endpoint}/{id}"

    def get(self, id: int):
        with self.lock:
            with httpx.Client() as client:
                response = client.get(self.get_url(id))
                if response.status_code == 200:
                    data = response.json()
                    obj = self.obj_class.parse_obj(data)
                    return obj
                else:
                    response.raise_for_status()

    def post(self, obj):
        with self.lock:
            with httpx.Client() as client:
                response = client.post(self.get_url(), json=obj.dict(by_alias=True), headers={'Content-Type': 'application/json'})
                if response.status_code == 201:
                    data = response.json()
                    obj = self.obj_class.parse_obj(data)
                    return obj
                else:
                    response.raise_for_status()

File: backend/Application/ApiClient.cs
using System.Text;
using System.Text.Json;
using Domain.Logic;

namespace site;

public class ApiClient<T> where T : class
{
    private readonly object @lock = new();
    private static readonly HttpClient HttpClient = new();
    private readonly string RequestUri;
    private static readonly JsonSerializerOptions OptionsDeserializer = new() { PropertyNameCaseInsensitive = true };

    public ApiClient()
    {
        var apiBaseUrl = HostsUrlGetter.ApiUrl;
        RequestUri = $"{apiBaseUrl}/api/{typeof(T).Name.ToLower()}";
    }

    public T? Get(int id)
    {
        lock (@lock)
        {
            var response = HttpClient.GetAsync($"{RequestUri}/{id}").Result;

            if (!response.IsSuccessStatusCode) return null;

            var jsonData = response.Content.ReadAsStringAsync().Result;
            var obj = JsonSerializer.Deserialize<T>(jsonData, OptionsDeserializer);
            return obj;
        }
    }

    public void Post(T obj)
    {
        lock (@lock)
        {
            var jsonContent = JsonSerializer.Serialize(obj);
            var content = new StringContent(jsonContent, Encoding.UTF8, "application/json");

            var response = HttpClient.PostAsync($"{RequestUri}", content).Result;
            if (response.IsSuccessStatusCode) return;

            throw new HttpRequestException($"Error posting data: {response.StatusCode}");
        }
    }

    public void Delete(int id)
    {
        lock (@lock)
        {
            var response = HttpClient.DeleteAsync($"{RequestUri}/{id}").Result;

            if (response.IsSuccessStatusCode) return;

            throw new HttpRequestException($"Error deleting data: {response.StatusCode}");
        }
    }
}